### ISSUE TITLE
KustoClient(kcsb): can only concatenate list (not "NoneType") to list

### ISSUE THREAD

**Issue by ecc256** [Azure/azure-kusto-python#304]
Getting bizarre error with KustoClient()
`can only concatenate list (not "NoneType") to list`
Same error for both `with_aad_application_key_authentication` and `with_aad_user_password_authentication`
and the same exact credentials work from C# app just fine.
C# connection string looks same with a couple extra settings:
```
Initial Catalog=NetDefaultDB;
dSTS Federated Security=False;
```
I’ve tried to append them to Python connection string, but it didn’t like them...
Could somebody point me to troubleshooting instructions, please?

python env:
```
conda list
azure-common              1.1.26                   pypi_0    pypi
azure-core                1.11.0                   pypi_0    pypi
azure-identity            1.5.0                    pypi_0    pypi
azure-kusto-data          2.0.0                    pypi_0    pypi
azure-kusto-ingest        2.0.0                    pypi_0    pypi
azure-mgmt-core           1.2.2                    pypi_0    pypi
azure-mgmt-datafactory    1.0.0                    pypi_0    pypi
azure-mgmt-kusto          1.0.0                    pypi_0    pypi
azure-mgmt-resource       16.0.0                   pypi_0    pypi
azure-storage-blob        12.8.0                   pypi_0    pypi
azure-storage-queue       12.1.5                   pypi_0    pypi
ca-certificates           2021.1.19            haa95532_0
certifi                   2020.12.5        py39haa95532_0
cffi                      1.14.5                   pypi_0    pypi
chardet                   4.0.0                    pypi_0    pypi
cryptography              3.4.6                    pypi_0    pypi
idna                      2.10                     pypi_0    pypi
isodate                   0.6.0                    pypi_0    pypi
msal                      1.9.0                    pypi_0    pypi
msal-extensions           0.3.0                    pypi_0    pypi
msrest                    0.6.21                   pypi_0    pypi
numpy                     1.20.1                   pypi_0    pypi
oauthlib                  3.1.0                    pypi_0    pypi
openssl                   1.1.1j               h2bbff1b_0
pandas                    1.2.3                    pypi_0    pypi
pip                       21.0.1           py39haa95532_0
portalocker               1.7.1                    pypi_0    pypi
pycparser                 2.20                     pypi_0    pypi
pyjwt                     2.0.1                    pypi_0    pypi
python                    3.9.1                h6244533_2
python-dateutil           2.8.1                    pypi_0    pypi
pytz                      2021.1                   pypi_0    pypi
pywin32                   300                      pypi_0    pypi
requests                  2.25.1                   pypi_0    pypi
requests-oauthlib         1.3.0                    pypi_0    pypi
setuptools                52.0.0           py39haa95532_0
sqlite                    3.33.0               h2a8f88b_0
tzdata                    2020f                h52ac0ba_0
urllib3                   1.26.3                   pypi_0    pypi
vc                        14.2                 h21ff451_1
vs2015_runtime            14.27.29016          h5e58377_2
wheel                     0.36.2             pyhd3eb1b0_0
wincertstore              0.2              py39h2bbff1b_0
zlib                      1.2.11               h62dcd97_4

```

**Comment by vladikbr** [comment]
Can you provide code sample (omitting any secrets, of course)? And the full stack trace for your error.

**Comment by ecc256** [comment]
```
from azure.kusto.data import KustoClient, KustoConnectionStringBuilder
import sys

client_id = '' # Application ID
client_secret = '' # Client Secret
authority_id = ''

cluster = 'name.eastus2'
# cluster = 'name'
database = 'name'

uri = f'https://{cluster}.kusto.windows.net/{database}'

kcsb = KustoConnectionStringBuilder.with_aad_application_key_authentication(uri, client_id, client_secret, authority_id)
print(sys.platform)
print(kcsb.data_source)
print(kcsb.aad_federated_security)
# kcsb.data_source = database
print(kcsb)

a = KustoClient(kcsb)
```

_win32
https://name.eastus2.kusto.windows.net/name
True
Data Source=https://name.eastus2.kusto.windows.net/name;AAD Federated Security=True;Application Client Id=;Application Key=****;Authority Id=_

last like throws:
**can only concatenate list (not "NoneType") to list**

**Comment by yihezkel** [comment]
Are you providing the authority_id? That's a required parameter.
With or without an authority_id, I'm unable to reproduce. I suspect the issue is related to different package versions.

To determine the issue, it'd be helpful to see the full stack trace.

**Comment by ecc256** [comment]
> Are you providing the authority_id? That's a required parameter.

Yep, but I use tenant_id instead...

> I suspect the issue is related to different package versions.

Could be, which ones do you use?

> To determine the issue, it'd be helpful to see the full stack trace.

How do I get it exactly?

**Comment by ecc256** [comment]
This [one](https://docs.python.org/3/library/traceback.html) will do?

**Comment by ecc256** [comment]
Is this good enough?
```
Traceback (most recent call last):
  File ".\query.py", line 101, in <module>
    a = KustoClient(kcsb)
  File ".\.venv\lib\site-packages\azure\kusto\data\client.py", line 606, in __init__
    socket_options=HTTPConnection.default_socket_options + self.compose_socket_options(), pool_maxsize=self._max_pool_size
TypeError: can only concatenate list (not "NoneType") to list
```
Does it mean, `self.compose_socket_options()` is `None`?

**Comment by yihezkel** [comment]
Yes, that stack trace helps, thanks!
This is a known issue when the code is executed in an environment where the OS isn't Windows/Linux/Mac or can't be identified. It's been fixed but not yet released. We'll work on getting a release out shortly.

**Comment by ecc256** [comment]
> This is a known issue when the code is executed in an environment where the OS isn't Windows/Linux/Mac or can't be identified.

Ok.
It's win32... Anyway to give a hint to `KustoClient`?

**Comment by yihezkel** [comment]
It's based on the "sys.platform" value.

**Comment by ecc256** [comment]
The [code above](https://github.com/Azure/azure-kusto-python/issues/304#issuecomment-800433254) run on **Windows Server 2016 Datacenter** 
`print(sys.platform)`
_win32_
Is it wrong value or it’s not recognized for some reason?

**Comment by yihezkel** [comment]
Apparently in that environment, one of the following are false:
- hasattr(socket, "SOL_SOCKET")
- hasattr(socket, "SO_KEEPALIVE")
- hasattr(socket, "TCP_KEEPIDLE")
- hasattr(socket, "TCP_KEEPCNT")

**Comment by yihezkel** [comment]
Resolved in version 2.1.0 (https://github.com/Azure/azure-kusto-python/releases/tag/2.1.0)

**Comment by ecc256** [comment]
@yihezkel 
I'm able to connect with 2.1.0
Thanks a lot for fixing it!!!

There is a minor issue:
This endpoint works:
`uri = f'https://{cluster}.kusto.windows.net/'`

But NOT this one:
`uri = f'https://{cluster}.kusto.windows.net/{database}'`
Is it but design?

Both work properly with C# SDK
